### PR TITLE
feat: pattern mining with drain

### PR DESCRIPTION
**Pull request by kolesnikovae**
A basic implementation of pattern mining with Drain. Persistence, durability, reliability, performance, and many other things are out of scope of the experiment.

How to test:
 1. Run [docker-compose](https://github.com/grafana/loki-hackathon-2024-03-logger-info/blob/feat/pattern-mining-with-drain/examples/getting-started/docker-compose.yaml).
 2.  Query patterns with curl: [./examples/getting-started/query.sh](https://github.com/grafana/loki-hackathon-2024-03-logger-info/blob/feat/pattern-mining-with-drain/examples/getting-started/query.sh). The output should be similar to:

```json
{
  "patterns": [
    {
      "name": "",
      "pattern": "192.168.144.7 - - <*> +0000] \"POST /loki/api/v1/push HTTP/1.1\" <*> <*> \"-\" \"promtail/2.9.2\"",
      "sampleLogLines": [
        "192.168.144.7 - - [01/Mar/2024:08:27:33 +0000] \"POST /loki/api/v1/push HTTP/1.1\" 204 0 \"-\" \"promtail/2.9.2\"",
        "192.168.144.7 - - [01/Mar/2024:08:27:34 +0000] \"POST /loki/api/v1/push HTTP/1.1\" 502 157 \"-\" \"promtail/2.9.2\"",
        "192.168.144.7 - - [01/Mar/2024:08:27:45 +0000] \"POST /loki/api/v1/push HTTP/1.1\" 499 0 \"-\" \"promtail/2.9.2\"",
        "192.168.144.7 - - [01/Mar/2024:08:27:47 +0000] \"POST /loki/api/v1/push HTTP/1.1\" 502 157 \"-\" \"promtail/2.9.2\"",
      ],
      "volumeTimeSeries": [
        [
          1709281650000000000,
          1
        ],
        [
          1709281660000000000,
          3
        ],
        ...
      ],
      "matches": 1003
    },
    {
      "name": "",
      "pattern": "2024/03/01 08:27:34 [error] 31#31: *1 connect() failed (111: Connection refused) while connecting to upstream, client: 192.168.144.7, server: , request: \"POST /loki/api/v1/push HTTP/1.1\", upstream: \"http://192.168.144.4:3100/loki/api/v1/push\", host: \"gateway:3100\"",
      "sampleLogLines": [
        "2024/03/01 08:27:34 [error] 31#31: *1 connect() failed (111: Connection refused) while connecting to upstream, client: 192.168.144.7, server: , request: \"POST /loki/api/v1/push HTTP/1.1\", upstream: \"http://192.168.144.4:3100/loki/api/v1/push\", host: \"gateway:3100\""
      ],
      "volumeTimeSeries": [
        [
          1709281654862583000,
          1
        ]
      ],
      "matches": 1
    },
    {
      "name": "",
      "pattern": "2024/03/01 <*> [error] 33#33: *266 loki could not be resolved (3: Host not found), client: 192.168.144.7, server: , request: \"POST /loki/api/v1/push HTTP/1.1\", host: \"gateway:3100\"",
      "sampleLogLines": [
        "2024/03/01 08:27:47 [error] 33#33: *266 loki could not be resolved (3: Host not found), client: 192.168.144.7, server: , request: \"POST /loki/api/v1/push HTTP/1.1\", host: \"gateway:3100\"",
        "2024/03/01 08:27:49 [error] 33#33: *266 loki could not be resolved (3: Host not found), client: 192.168.144.7, server: , request: \"POST /loki/api/v1/push HTTP/1.1\", host: \"gateway:3100\"",
      ],
      "volumeTimeSeries": [
        [
          1709281660000000000,
          1
        ],
        ...
      ],
      "matches": 3
    }
  ]
}
```

Note that, as of now, the backend requires a query to target a specific service (e.g., `{"service_name"="my-service"}`.)

TODO:
 - [ ] Structured logs: JSON and logfmt – need a better tokenizer
 - [ ] Make LogQL to filter by pattern
 - [ ] Pattern naming
 - [ ] Field classification (named entity recognition, type detection, etc)

/cc @cyriltovena

### CODE REVIEW

File: pkg/loggerinfo/drain/drain.go [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+	hi := len(x.Values)
+	if end < last {
+		hi = sort.Search(len(x.Values), func(i int) bool {
+			return x.Values[i][0] >= end
+		})
+	}
+	return &Volume{
+		Values: x.Values[lo:hi],
+	}
+}
```
Comment (kolesnikovae) [inline pkg/loggerinfo/drain/drain.go:110]: @cyriltovena, I think there is a bug: the time series range start and end do not match the query, but the data we have – I will fix that

**Comment by cyriltovena** [comment]
I will build it like this to test the API for now.

File: examples/getting-started/docker-compose.yaml [hunk 2]
```diff
@@ -1,34 +1,17 @@
...
-    networks: &loki-dns
-      loki:
-        aliases:
-          - loki
-
-  write:
-    image: grafana/loki:2.9.2
-    command: "-config.file=/etc/loki/config.yaml -target=write"
+  loki:
+    image: kolesnikovae/loki-loggerinfo:latest
```
Comment (cyriltovena) [inline examples/getting-started/docker-compose.yaml:9]: ho you did already built it. Let me use that.

**Comment by cyriltovena** [comment]
Going to also change the API because I think we want get style one like other API

**Comment by cyriltovena** [comment]
I faced this too 

```
fatal error: concurrent map read and map write

goroutine 1220 [running]:
github.com/hashicorp/golang-lru/simplelru.(*LRU).Get(0xc002811880, {0xb09ccc0, 0xca626c8})
````

**Pull request merged**